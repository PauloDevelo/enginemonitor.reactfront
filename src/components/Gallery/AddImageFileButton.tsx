import React from 'react';
import {Button} from 'reactstrap';

import { faImage, faPlus } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { ImageModel } from '../../types/Types';

import {resizeAndSaveImage} from '../../helpers/ImageHelper';

import errorService from '../../services/ErrorService';

type Props = {
    parentUiId: string,
    className?: string,
    addImage: (image: ImageModel) => void
}

const createFileSelector = (onChange: (e: Event)=>void):HTMLInputElement => {
    const fileSelector = document.createElement('input');
    fileSelector.setAttribute('type', 'file');
    fileSelector.addEventListener('change', onChange);

    return fileSelector;
}

function AddImageFileButton({parentUiId, addImage, className}: Props){
    const onFileSelected = (file: File) => {
        uploadImageFile(file, "multer");
    };

    const uploadImageFile = async(file: File, method: string) => {
        try{
            let newImage:ImageModel | undefined = undefined;

            if(method === "multer"){
                newImage = await resizeAndSaveImage(file, parentUiId);
                addImage(newImage);
            }
        }
        catch(error){
            errorService.addError(error);
        }
    };

    const onChange = (e: Event) => {
        const changeEvent = e as unknown as React.ChangeEvent<HTMLInputElement>;
        const files = changeEvent.target.files ? changeEvent.target.files : new FileList() ;
        
        if(files.length > 0){
            onFileSelected(files[0]);
        }
    }

    const fileSelector = createFileSelector(onChange);

    const handleFileSelect = (e: React.MouseEvent):void => {
        e.preventDefault();
        fileSelector.click();
    }

    return (
    <Button color="light" size="lg" className={className} onClick={handleFileSelect} aria-label="Edit">
        <span className="fa-layers fa-fw">
            <FontAwesomeIcon icon={faImage} size="lg"/>
            <FontAwesomeIcon icon={faPlus} size="xs" transform="down-13 left-16"/>
        </span>
    </Button>);
}

export default React.memo(AddImageFileButton);