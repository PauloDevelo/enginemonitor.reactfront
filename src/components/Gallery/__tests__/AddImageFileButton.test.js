import React from 'react';
import { mount } from 'enzyme';

import AddImageFileButton from '../AddImageFileButton';

describe('Component Thumbnail', () => {
  it('should render an button', () => {
    // Arrange
    const onImageAdded = jest.fn();

    // Act
    const buttonAddImage = mount(<AddImageFileButton parentUiId="parent_01" addImage={onImageAdded} />);

    // Assert
    expect(buttonAddImage).toMatchSnapshot();
  });
});
