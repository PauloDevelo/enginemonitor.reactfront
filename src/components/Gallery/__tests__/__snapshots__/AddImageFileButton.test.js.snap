// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Thumbnail should render an button 1`] = `
<Memo(AddImageFileButton)
  addImage={[MockFunction]}
  parentUiId="parent_01"
>
  <Button
    aria-label="Edit"
    color="light"
    onClick={[Function]}
    tag="button"
  >
    <button
      aria-label="Edit"
      className="btn btn-light"
      onClick={[Function]}
      type="button"
    >
      <span
        className="fa-layers fa-fw"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f03e",
                "M464 448H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h416c26.51 0 48 21.49 48 48v288c0 26.51-21.49 48-48 48zM112 120c-30.928 0-56 25.072-56 56s25.072 56 56 56 56-25.072 56-56-25.072-56-56-56zM64 384h384V272l-87.515-87.515c-4.686-4.686-12.284-4.686-16.971 0L208 320l-55.515-55.515c-4.686-4.686-12.284-4.686-16.971 0L64 336v48z",
              ],
              "iconName": "image",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size="lg"
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-image fa-w-16 fa-lg "
            data-icon="image"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 448H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h416c26.51 0 48 21.49 48 48v288c0 26.51-21.49 48-48 48zM112 120c-30.928 0-56 25.072-56 56s25.072 56 56 56 56-25.072 56-56-25.072-56-56-56zM64 384h384V272l-87.515-87.515c-4.686-4.686-12.284-4.686-16.971 0L208 320l-55.515-55.515c-4.686-4.686-12.284-4.686-16.971 0L64 336v48z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f067",
                "M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z",
              ],
              "iconName": "plus",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size="xs"
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform="down-13 left-16"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-plus fa-w-14 fa-xs "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={
              Object {
                "transformOrigin": "-0.5625em 1.3125em",
              }
            }
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              style={Object {}}
              transform="translate(224 256)"
            >
              <g
                style={Object {}}
                transform="translate(-512, 416)  scale(1, 1)  rotate(0 0 0)"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                  style={Object {}}
                  transform="translate(-224 -256)"
                />
              </g>
            </g>
          </svg>
        </FontAwesomeIcon>
      </span>
    </button>
  </Button>
</Memo(AddImageFileButton)>
`;
